### PR TITLE
Add tests for passwdqc pwqcheck

### PR DESCRIPTION
**Pull request by Zaiba-S**
Currently, there are no tests implemented for the package, so this serves as an initial attempt to implement basic testing coverage.
This PR introduces an initial set of tests for the pwqcheck binary of the passwdqc package. 

Any feedback or suggestions would be appreciated.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -307,6 +307,17 @@ remove_lib_wrapped:
 remove_locales_wrapped:
 	for f in $(LANGUAGES); do $(RM) $(DESTDIR)$(LOCALEDIR)/$$f/LC_MESSAGES/$(PACKAGE).mo; done
 
+test:
```
Comment (solardiz) [inline Makefile:310]: I think we want this target called `check`, and it should exit non-zero on any test failures.

And we should run it from GitHub Actions, but that may be a separate PR or at least separate commit.

File: Makefile [hunk 2]
```diff
@@ -307,6 +307,17 @@ remove_lib_wrapped:
...
+	@echo "Running tests..."
+	# Set LD_LIBRARY_PATH to include the current directory
+	@export LDFLAGS="-Wl,-rpath,$(CURDIR)"; \
+	export LD_LIBRARY_PATH=$(CURDIR):$$LD_LIBRARY_PATH; \
```
Comment (solardiz) [inline Makefile:314]: I think this is excessive. For my manual testing, I simply set `LD_LIBRARY_PATH=.` and it just works.

The change of `LDFLAGS` wouldn't take effect at this point anyway, because the test scripts being invoked from here assume the tool has already been built.

**Review by solardiz — COMMENTED** [review]
Thank you for the contribution. Yes, we need tests.

**Comment by solardiz** [comment]
My thinking was to start with what I've been doing in manual testing - running pwqcheck on a large file and confirming the hash of its output isn't unexpectedly changed. This wouldn't directly test the individual checks, but it would ensure stable behavior. So if we ever make code changes that change the behavior on such large test input, we'll need to explicitly acknowledge this and update the hash. A drawback is the large input file would need to be stored somewhere in here or download from somewhere external (e.g., the `run/password.lst` file we have in the `john` repo) or reproducibly generated.

```
$ LD_LIBRARY_PATH=. ./pwqcheck -1 --multi < ~/john/run/password.lst | grep -c ^OK:
1733
$ LD_LIBRARY_PATH=. ./pwqcheck -1 --multi < ~/john/run/password.lst | grep -c ^Bad
1793975
$ LD_LIBRARY_PATH=. ./pwqcheck -1 --multi < ~/john/run/password.lst | md5sum 
3eccb6ef49eb84b494c61954c240e76c  -
```

I see that you also test a few features that the above would not.

File: tests/test-pwqcheck-basic.sh [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+test_basic_password "MyP@ssw0rd" "pass" "Mixed case with symbols and numbers"
+test_basic_password "Str0ng!P@ssphrase" "pass" "Strong with multiple special chars"
+test_basic_password "C0mpl3x1ty!" "pass" "Complex but reasonable length"
+
+# Test Suite 4: Edge Cases
+echo -e "\nTesting Edge Cases:"
+test_basic_password " " "fail" "Single space"
+test_basic_password "" "fail" "Empty password"
+test_basic_password "$(printf 'a%.0s' {1..71})" "fail" "Very long password"
+test_basic_password "ljy8zk9aBJ3hA3TXAAMAQe61ytFohJM4SuPFbA4L1xDqV2JDE1n8BCnLN96evcJMWyTkr9y3" "pass" "Max lenght password"
```
Comment (solardiz) [inline tests/test-pwqcheck-basic.sh:60]: Typo: `lenght`.

Also, if we do these tests, there should probably be one for a password that's just above max length, with expected `fail`.

Also, shouldn't we expect specific `fail` reasons?

Comment (Zaiba-S) [reply tests/test-pwqcheck-basic.sh:60]: Thanks for the review !

-  I will correct the typo and add test case for a password just above the max length and ensure it fails as expected.

-  Reagarding the fail reason,Could you please clarify if you are expecting specific fail reasons for each test case?

**Comment by Zaiba-S** [comment]
> My thinking was to start with what I've been doing in manual testing - running pwqcheck on a large file and confirming the hash of its output isn't unexpectedly changed. This wouldn't directly test the individual checks, but it would ensure stable behavior. So if we ever make code changes that change the behavior on such large test input, we'll need to explicitly acknowledge this and update the hash. A drawback is the large input file would need to be stored somewhere in here or download from somewhere external (e.g., the `run/password.lst` file we have in the `john` repo) or reproducibly generated.
> 
> ```
> $ LD_LIBRARY_PATH=. ./pwqcheck -1 --multi < ~/john/run/password.lst | grep -c ^OK:
> 1733
> $ LD_LIBRARY_PATH=. ./pwqcheck -1 --multi < ~/john/run/password.lst | grep -c ^Bad
> 1793975
> $ LD_LIBRARY_PATH=. ./pwqcheck -1 --multi < ~/john/run/password.lst | md5sum 
> 3eccb6ef49eb84b494c61954c240e76c  -
> ```
> 
> I see that you also test a few features that the above would not.

Thanks for sharing your approach 
Your method of hashing the output of pwqcheck on a large input file sounds like a great way to ensure stability across changes. 
Are you suggesting that we implement this approach in addition to the individual tests I proposed? 
If so, For storing the large input file and expected output hash for comparison, do you have a preference on where it should be stored ?

**Comment by solardiz** [comment]
This PR currently fails our whitespace-errors test:

```

Run git diff-index --check --cached 4b825dc642cb6eb9a060e54bf8d69288fbee4904
tests/test-pwqcheck-length.sh:14: trailing whitespace.
+    
tests/test-pwqcheck-length.sh:18: trailing whitespace.
+    
tests/test-pwqcheck-length.sh:22: trailing whitespace.
+    
tests/test-pwqcheck-multi.sh:13: trailing whitespace.
+    
tests/test-pwqcheck-multi.sh:16: trailing whitespace.
+    
tests/test-pwqcheck-multi.sh:25: trailing whitespace.
+    
tests/test-pwqcheck-multi.sh:39: trailing whitespace.
+    
tests/test-pwqcheck-multi.sh:85: trailing whitespace.
+  
tests/test-pwqcheck-multi.sh:109: trailing whitespace.
+            echo "StrongP@ss${i}!" 
tests/test-pwqcheck-multi.sh:111: trailing whitespace.
+            echo "weak${i}" 
tests/test-pwqcheck-similarity.sh:94: new blank line at EOF.
```

**Comment by solardiz** [comment]
> Are you suggesting that we implement this approach in addition to the individual tests I proposed?

In general, yes. In this same PR, no.

> If so, For storing the large input file and expected output hash for comparison, do you have a preference on where it should be stored ?

Currently no. I'd like to hear from @ldv-alt on this. Off the top of my head, the options are:

1. In this repo and included in release archive (so that `make check` can use it from there as well).
2. In this repo, but excluded from release archive (available for use in GitHub Actions).
3. External reference (what kind?) to `john/run/password.lst`.
4. Generate a large list into a pipe programmatically (the lines wouldn't look so much like real passwords, but that's OK if we're only testing stability).

File: tests/test-pwqcheck-basic.sh (hunk 3 above)
Comment (solardiz) [reply tests/test-pwqcheck-basic.sh:60]: > * Reagarding the fail reason,Could you please clarify if you are expecting specific fail reasons for each test case?

Yes, like you already have in another script where you check against expected output. Instead of the word `fail`, you could include those reason strings.

**Comment by solardiz** [comment]
The color output could be undesired when redirected to a build log file. I suggest to use color only when outputting to a tty, which I think you can test with `test -t 1`, e.g. `if [ -t 1 ]; then`.

We use 8-wide tabs for indentation in this source tree, and I think it makes sense to continue with this style for the shell scripts in here as well (not switch to 4 spaces).

File: tests/test-pwqcheck-basic.sh [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    local description="$3"
+
+    PWQCHECK_BIN="$(dirname "$0")/../pwqcheck"
+
+    echo -n "Testing $description: "
+    result=$(echo "$password" | "$PWQCHECK_BIN" -1 2>&1)
+    exit_code=$?
+
+    if [[ "$expected" == "pass" && $exit_code -eq 0 ]] || \
+       [[ "$expected" == "fail" && $exit_code -ne 0 ]]; then
```
Comment (solardiz) [inline tests/test-pwqcheck-basic.sh:21]: This could be more portable (to other shells) if written as a `test` expression (so in single brackets and using `-a` and `-o` for AND and OR) rather than as two `bash` conditional expressions.

**Comment by solardiz** [comment]
Please also add your copyright statements and references to `LICENSE` as source code comments near the start of your scripts, similarly to how we have this in existing source files.

**Comment by solardiz** [comment]
> This PR currently fails our whitespace-errors test:
> 
> ```
> Run git diff-index --check --cached 4b825dc642cb6eb9a060e54bf8d69288fbee4904
> ```

This is still failing. You can run the above `git` command on your computer to see (and then fix) these errors before you push.

File: tests/test-pwqcheck-basic.sh (hunk 4 above)
Comment (Zaiba-S) [reply tests/test-pwqcheck-basic.sh]: should we set the default shell to sh instead of bash?

Comment (solardiz) [reply tests/test-pwqcheck-basic.sh]: We could, if we actually test these scripts with non-bash as well.
